### DISCUSSION TITLE
78. 인덱스에서 사용되는 자료 구조에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#80] [General]
- B-Tree와 B+ Tree의 차이점에 대해 설명해 주세요.
- 해시 인덱스에서 범위검색과 순차검색을 지원하지 않는 이유는 무엇인가요?

**Comment by linirini** [comment]
- 해시 테이블
  - 키를 해시 함수를 통해 계산하여 인덱스 위치에 데이터를 저장하는 자료구조
  - 데이터에 대한 빠른 접근을 가능하게 하며, 평균적으로 O(1)의 시간 복잡도로 데이터를 검색
  - 하지만 해시 테이블은 순서가 없기 때문에 범위 쿼리나 정렬된 데이터를 필요로 하는 경우에는 적합하지 않다.

- B-트리
  -  데이터베이스와 파일 시스템에서 자주 사용되는 균형 트리 구조
  - 노드가 여러 자식을 가질 수 있어 높은 차수의 트리를 형성하고, 각 노드는 정렬된 키를 포함
  - B-트리는 데이터를 검색, 삽입, 삭제할 때 O(log n)의 시간 복잡도를 제공하며, 디스크 페이지를 효율적으로 사용할 수 있도록 설계됨

- B+트리
  - B+트리는 B-트리의 변형으로, 모든 값이 리프 노드에만 저장되고 내부 노드는 검색을 위한 키만 보유하는 구조
  - 리프 노드들 사이에 포인터를 연결하여 범위 쿼리를 효율적으로 처리
  - B+트리는 디스크 I/O 작업을 최소화하여 성능을 최적화하며, B-트리와 마찬가지로 O(log n)의 시간 복잡도를 가짐

- B-Tree VS B+Tree
  - 구조 차이
    - B-Tree는 데이터와 키 모두 내부 노드와 리프 노드에 저장
    - B+Tree는 내부 노드에는 키만 저장되고, 데이터는 리프 노드에 저장되며, 리프 노드끼리는 연결리스트로 연결되어 있음
  - 순차 검색
    - B-Tree는 범위 검색은 가능하지만, 리프 노드끼리 연결되어 있지 않으므로 순차 검색은 어려움
    - B+Tree는 리프 노드가 연결리스트로 연결되어 있으므로 범위검색과 순차 검색에 효율적

- 해시 인덱스에서 범위 검색과 순차 검색이 지원되지 않는 이유
  - 해시 값은 특정 키와 매핑된 위치를 빠르게 찾을 수 있게 하지만, 해시 값 자체는 키 간 순서를 보장하지 않는다. 따라서 순차 검색은 어려움

**Comment by jongmee** [comment]
오타 발견 !! `균형 이진 트리 구조` -> `균형 트리 구조`
